dartagnan/src/main/java/com/dat3m/dartagnan/wmm/processing/FlattenAssociatives.java

```diff
@@ -0,0 +1,59 @@
+package com.dat3m.dartagnan.wmm.processing;
+
+import com.dat3m.dartagnan.wmm.Constraint;
+import com.dat3m.dartagnan.wmm.Definition;
+import com.dat3m.dartagnan.wmm.Relation;
+import com.dat3m.dartagnan.wmm.Wmm;
+import com.dat3m.dartagnan.wmm.definition.Intersection;
+import com.dat3m.dartagnan.wmm.definition.Union;
+
+import java.util.List;
+import java.util.function.BiFunction;
+import java.util.stream.Stream;
+
+// Flattens nested binary unions/intersections into n-ary ones.
+public class FlattenAssociatives implements WmmProcessor {
+
+    private FlattenAssociatives() {
+    }
+
+    public static FlattenAssociatives newInstance() {
+        return new FlattenAssociatives();
+    }
+
+    @Override
+    public void run(Wmm wmm) {
+        flattenAssociatives(wmm, Union.class, Union::new);
+        flattenAssociatives(wmm, Intersection.class, Intersection::new);
+    }
+
+    private void flattenAssociatives(Wmm wmm,
+                                     Class<? extends Definition> cls,
+                                     BiFunction<Relation, Relation[], Definition> constructor
+    ) {
+        final List<Relation> relations = List.copyOf(wmm.getRelations());
+        final List<Constraint> constraints = wmm.getConstraints().stream()
+                .filter(c -> !(c instanceof Definition)).toList();
```

Comment:
Wouldn't make more sense to use `c instanceof Constraint` for the filtering? If at some point `wmm.getConstraints()` could return yet a 3rd class, the current implementation would not have the intended behavior.

Reply:
`c instanceof Constraint` is true for everything returned by `w..getConstraints`, so no, it would not make more sense :P. The intention here is to get all constraints that are no definitions. 
That being said, I have not taken a proper look into this pass; this is preexisting code that was just moved from `Wmm` to an isolated pass.